x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/aggs/heuristic/PValueScoreTests.java

```diff
@@ -0,0 +1,146 @@
+        return true;
+    }
+
+    @Override
+    public void testAssertions() {
+        testBackgroundAssertions(new PValueScore(true), new PValueScore(false));
+    }
+
+    @Override
+    protected NamedXContentRegistry xContentRegistry() {
+        return new NamedXContentRegistry(
+            new SearchModule(Settings.EMPTY, Arrays.asList(new MachineLearning(Settings.EMPTY, null))).getNamedXContents()
+        );
+    }
+
+    @Override
+    protected NamedWriteableRegistry writableRegistry() {
+        return new NamedWriteableRegistry(
+            new SearchModule(Settings.EMPTY, Arrays.asList(new MachineLearning(Settings.EMPTY, null))).getNamedWriteables()
+        );
+    }
+
+    public void testPValueScore_WhenAllDocsContainTerm() {
+        PValueScore pValueScore = new PValueScore(randomBoolean());
+        long supersetCount = randomNonNegativeLong();
+        long subsetCount = randomLongBetween(0L, supersetCount);
+        assertThat(pValueScore.getScore(subsetCount, subsetCount, supersetCount, supersetCount), equalTo(0.0));
+    }
+
+    public void testHighPValueScore() {
+        boolean backgroundIsSuperset = randomBoolean();
+        long supersetCount = randomLongBetween(0L, Long.MAX_VALUE/2);
+        long subsetCount = randomLongBetween(0L, supersetCount);
+        if (backgroundIsSuperset) {
+            supersetCount += subsetCount;
+        }
+
+        PValueScore pValueScore = new PValueScore(backgroundIsSuperset);
+        assertThat(pValueScore.getScore(subsetCount, subsetCount, subsetCount, supersetCount), greaterThanOrEqualTo(700.0));
+    }
+
+    public void testLowPValueScore() {
+        boolean backgroundIsSuperset = randomBoolean();
+        long supersetCount = randomLongBetween(0L, Long.MAX_VALUE/2);
+        long subsetCount = randomLongBetween(0L, supersetCount);
+        long subsetFreqCount = randomLongBetween(0L, subsetCount/5);
+        if (backgroundIsSuperset) {
+            supersetCount += subsetCount;
+        }
+
+        PValueScore pValueScore = new PValueScore(backgroundIsSuperset);
+        assertThat(
+            pValueScore.getScore(subsetFreqCount, subsetCount, subsetCount, supersetCount),
+            allOf(lessThanOrEqualTo(5.0), greaterThanOrEqualTo(0.0))
+        );
+    }
+
+    public void testPValueScore() {
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 100, 1000)),
+            closeTo(1.0, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 10, 1000)),
+            closeTo(0.0014942974424670364, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 200, 1000)),
+            closeTo(1.0, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(20, 10000, 5, 10000)),
+            closeTo(0.3154715149153073, eps)
+        );
+    }
+
+    public void testSmallChanges() {
```

Comment:
👍

Reply:
I would also add tests that similar ratios and very large counts produce largish p-values, e.g. things like
```
getScore(11000, 100000, 50000, 500000)
getScore(12000, 100000, 50000, 500000)
getScore(13000, 100000, 50000, 500000)
```